test/unit/test_json.c

```diff
@@ -0,0 +1,338 @@
+		"\xd1\x82\xd0\xb5\xd0\xba\xd1\x81\xd1\x82\"\n  ]\n}\n"
+	}, { // 28
+		"2009-10-20@20:38:21.539575\n",
+		"2009\n"
+	}, { // 29
+		"{ \"U+10ABCD\": \"\xf4\x8a\xaf\x8d\" }\n\n",
+		"{\n  U+10ABCD:\"\xf4\x8a\xaf\x8d\"\n}\n"
+	}, { // 30
+		"[ 1,2,3,4,5,6,7,\n  123456789 , -123456789,\n  2147483647, -2147483647"
+		" ]\n",
+		"[\n  1\n  2\n  3\n  4\n  5\n  6\n  7\n  123456789\n  -123456789\n  214"
+		"7483647\n  -2147483647\n]\n"
+	}, { // 31
+		"/* invalid utf-8 */ [\"\xd0\x94\xd0\xb0 \xd0\x9c\xd1 \xd0\x95\xd0\xb1"
+		"\xd0\xb0 \xd0\x9c\xd0\xb0\xd0\xb9\xd0\xba\xd0\xb0\xd1\x82\xd0\xb0\"]\n",
+		"[\n  \"\xd0\x94\xd0\xb0 \xd0\x9c\xd1 \xd0\x95\xd0\xb1\xd0\xb0 \xd0\x9c"
+		"\xd0\xb0\xd0\xb9\xd0\xba\xd0\xb0\xd1\x82\xd0\xb0\"\n]\n"
+	}, { // 32
+		"{ \"bad thing\": 010 }\n",
+		"{\n  bad thing:8\n}\n"
+	}, { // 33
+		"[\n\t\t\"foo\", true,\n\t\ttrue, \"blue\",\n\t\t\"baby where are you?"
+		"\", \"oh boo hoo!\",\n        - \n]\n\n",
+		NULL
+	}, { // 34
+		"}\n",
+		NULL
+	}, { // 35
+		"10.e2\n",
+		"1.000000e+03\n"
+	}, { // 36
+		"10e\n",
+		"1.000000e+01\n"
+	}, { // 37
+		"{\n\t\"boolean, true\": true,\n\t\"boolean, false\": false,\n\t\"null"
+		"\": null\n}\n",
+		"{\n  boolean, true:T\n  boolean, false:F\n  null:null\n}\n"
+	}, { // 38
+		"[\"\\n foo \\/ bar \\r\\f\\\\\\uffff\\t\\b\\\"\\\\ and you can't escap"
+		"e thi\\s\"]\n",
+		"[\n  \"\n foo / bar \x0d\x0c\\\xef\xbf\xbf\t\x08\"\\ and you can't esc"
+		"ape thi\\s\"\n]\n"
+	}, { // 39
+		"[\"\\n foo \\/ bar \\r\\f\\\\\\uffff\\t\\b\\\"\\\\\",\n \"\\\"and this"
+		" string has an escape at the beginning\",\n \"and this string has no e"
+		"scapes\" ]\n",
+		"[\n  \"\n foo / bar \x0d\x0c\\\xef\xbf\xbf\t\x08\"\\\"\n  \"\"and this"
+		" string has an escape at the beginning\"\n  \"and this string has no e"
+		"scapes\"\n]\n"
+	}, { // 40
+		"\"la di dah.  this is a string, and I can do this, \\n, but not this\n"
+		"\"\n",
+		"\"la di dah.  this is a string, and I can do this, \n, but not this\n"
+		"\"\n"
+	}, { // 41
+		"/*",
+		NULL
+	}, { // 42
+		"/*abc /",
+		NULL
+	}, { // 43
+		"/*/\\*/ 123",
+		"123\n"
+	}, { // 44
+		"/*/ 123",
+		NULL
+	}, { // 45
+		"[\"\\u0414\\u0430\",\n \"\\u041c\\u0443\",\n \"\\u0415\\u0431\\u0430\""
+		",\n \"\\u041c\\u0430\\u0439\\u043a\\u0430\\u0442\\u0430\"]\n",
+		"[\n  \"\xd0\x94\xd0\xb0\"\n  \"\xd0\x9c\xd1\x83\"\n  \"\xd0\x95\xd0"
+		"\xb1\xd0\xb0\"\n  \"\xd0\x9c\xd0\xb0\xd0\xb9\xd0\xba\xd0\xb0\xd1\x82"
+		"\xd0\xb0\"\n]\n"
+	}, { // 46
+		"\"\\u0066\\u006f\\u006f\\u0062\\u0061\\u0072\"\n",
+		"\"foobar\"\n"
+	}, { // 47
+		"/* invalid unicode surrogate */ \"\\ud800\"\n",
+		NULL
+	}, { // 48
+		"\"\\u041F\\u0440\\u043E\\u0432\\u0435\\u0440\\u043a\\u0430\"",
+		"\"\xd0\x9f\xd1\x80\xd0\xbe\xd0\xb2\xd0\xb5\xd1\x80\xd0\xba\xd0\xb0\"\n"
+	}, { // 49
+		"\"\\u04FG\"",
+		NULL
+	}, { // 50
+		"\"\\u0",
+		NULL
+	}, { // 51
+		"\"\\u04\"",
+		NULL
+	}
+};
+
+static void dump(const RJson *json, RStrBuf *out, int indent) {
```

Comment:
Make it public api as char *r_json_tostring(RJson)

Reply:
This isn't actually producing json, it's just some random format used for comparing test results, so not really useful in the API. For testing this format actually makes more sense than printing actual json because you can compare possibly invalid utf-8 strings on the binary level.

Reply:
Would be better to stick to json. Maybe using the pj api

Reply:
As said for this special case the raw dumping is slightly better and has less behavorial dependencies. But a printing for actual usage could easily be built with pj. I'm just not sure if it is really practically useful.

Reply:
What would you say about instead of comparing these dumped strings at all, doing it all directly in C by `mu_assert(...` calls? I could probably write a code generator for all these cases so I don't have to write them from scratch. If we would go through pj I think we would test too much interdependent code here when we just want to test the parser and nothing else.

Reply:
Removed the non-standard dumping stuff and replaced with direct tests.